joinmarket-webui/jam
# feat: spend fidelity bond

Partly addresses #547

First iteration of moving a fidelity bond after it expired.
Feedback is highly welcome, especially in regard to wording.

A follow-up PR for "renewing" the FB is in the making (also part of #547).

First attempts of this feature were based on the "Create FB" flow:
- Select target jar
- Freeze non-participating UTXOs manually
- Unfreeze the Fidelity Bond UTXO
- Display payment info and confirm
- Send FB UTXO via direct-send
- Restore state of previously frozen UTXOs


While this method gives users a better understanding of what is going on, it was annoying, resulted in so much code and just felt bloated.

Now it is just:
- Select target jar
- Display payment info and confirm

With the current approach, the necessary intermediate steps are hidden from the user. This makes for a much nicer UX. Any input on this is highly appreciated.

Also, the resulting method has been generalized: Sending any amount of specific UTXOs to a generic destination (internal or external). This can be reused in the upcoming PR when renewing the FB and also elsewhere.

## Other stuff
- disallow unfreezing of expired fidelity bonds from `JarDetailsOverlay` component
  - this is necessary as unfrozen, expired FB cannot take part in taker operations (when starting a maker operation they are auto-frozen by JM, but unfortunately not for taker operations)
  - Fidelity Bond UTXOs will always show the lock icon
- externalized `PaymentConfirmModal` to be used from other components than `Send`
- prepare and lay ground for #252 to improve error messages coming from the backend
- order fidelity bond UTXOs by value and locked state
- prevent `Balance` component from unnecessary re-renders (+handle possible `NaN` values)

**Note**: Spending an expired fidelity bond UTXO can still fail if the median time of the chain is still lower than the locktime of the bond. There is currently no way to verify if it is successfully spendable and a bond will always be marked as "expired" if its locktime is in the past.

## How to test?

Create an expired fidelity bond (only possible when starting via `npm run dev:start`) and try moving it into a jar.
Also note, that only the actual fidelity bond UTXO should be spent, all other UTXOs should be frozen and not be moved. Once the payment is done (regardless of success), UTXOs should have their original state again.

## :camera_flash: 
<img src="https://user-images.githubusercontent.com/3358649/199501898-a562eab9-b13b-4c60-91de-efc23ca08225.png" width=300 /> <img src="https://user-images.githubusercontent.com/3358649/199501744-30b1ef23-d677-4c59-9bb4-985c3417351f.png" width=300 />

<img src="https://user-images.githubusercontent.com/3358649/199300809-90aed0a7-71b0-4dd9-9bec-bffd99df8c42.png" width=300 />
<img src="https://user-images.githubusercontent.com/3358649/199302359-19f7e778-2bfe-4e3d-b74a-4b8890c195ec.png" width=300 />
<img src="https://user-images.githubusercontent.com/3358649/199302452-38943ac1-3bf0-4c22-93b9-1b05173210d6.png" width=300 />
<img src="https://user-images.githubusercontent.com/3358649/199302489-a9e6851e-f29c-4a24-b521-d3d71051f704.png" width=300 />
